File: 0146-lru-cache/0146-lru-cache.java
class Node {
    int key;
    int val;
    Node next;
    Node prev;
    Node (int key, int val) {
        this.key = key;
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}

class LRUCache {
    HashMap<Integer, Node> hm;
    int size;
    Node left;
    Node right;

    public LRUCache(int capacity) {
        hm = new HashMap<Integer, Node>();
        size = capacity;
        left = new Node(0,0);
        right = new Node(0,0);
        left.next = right;
        right.prev = left;
    }
    
    public void remove(Node curr) {
        Node prev = curr.prev;
        Node next = curr.next;
        prev.next = next;
        next.prev = prev;
    }
    
    public void insert(Node curr) {
        Node prev = right.prev;
        prev.next = curr;
        curr.prev = prev;
        curr.next = right;
        right.prev = curr;
    }
    
    public int get(int key) {
        if (hm.containsKey(key))
        {
            //System.out.println(key);
            remove(hm.get(key));
            insert(hm.get(key));
            return hm.get(key).val;
        }
        return -1;
    }
    
    public void put(int key, int value) {
        if (hm.containsKey(key))
        {
            remove(hm.get(key));
        }
        Node newNod = new Node(key, value);
        hm.put(key, newNod);
        insert(newNod);
        //System.out.println(key + " " + hm.size());
        if (hm.size() > size)
        {
            Node lru = left.next;
            remove(lru);
            hm.remove(lru.key);
        }
        //System.out.println(key + " " + hm.size() + " g");
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class LRUCache:

    def __init__(self, capacity: int):
        self.queue = deque()
        self.capacity = capacity
        self.hashmap = defaultdict(int)

    def get(self, key: int) -> int:
        if key in self.hashmap:
            self.queue.remove(key)
            self.queue.append(key)
            return self.hashmap[key]
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.hashmap:
            temp = self.get(key)
        else:
            if len(self.queue) == self.capacity:
                temp = self.queue.popleft()
                del self.hashmap[temp]
            self.queue.append(key)
        self.hashmap[key] = value
        # print(self.queue, self.hashmap)



# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)